dataflow/gpu-examples/conftest.py

```diff
@@ -0,0 +1,508 @@
+        # due to incompatible dependencies with apache-beam.
+        # We use gcloud instead to cancel the job.
+        # https://cloud.google.com/sdk/gcloud/reference/dataflow/jobs/drain
+        cmd = [
+            "gcloud",
+            f"--project={project}",
+            "dataflow",
+            "jobs",
+            "drain",
+            job_id,
+            f"--region={region}",
+        ]
+        logging.info(f"{cmd}")
+        subprocess.run(cmd, check=True)
+
+        # https://cloud.google.com/sdk/gcloud/reference/dataflow/jobs/cancel
+        cmd = [
+            "gcloud",
+            f"--project={project}",
+            "dataflow",
+            "jobs",
+            "cancel",
+            job_id,
+            f"--region={region}",
+        ]
+        logging.info(f"{cmd}")
+        subprocess.run(cmd, check=True)
+        logging.info(f"Cancelled Dataflow job: {job_id}")
+
+    @staticmethod
+    def dataflow_flex_template_build(
+        bucket_name: str,
+        image_name: str,
+        metadata_file: str,
+        project: str = PROJECT,
+        template_file: str = "template.json",
+    ) -> str:
+        # https://cloud.google.com/sdk/gcloud/reference/dataflow/flex-template/build
+        template_gcs_path = f"gs://{bucket_name}/{template_file}"
+        cmd = [
+            "gcloud",
+            "dataflow",
+            "flex-template",
+            "build",
+            template_gcs_path,
+            f"--project={project}",
+            f"--image=gcr.io/{project}/{image_name}",
+            "--sdk-language=PYTHON",
+            f"--metadata-file={metadata_file}",
+        ]
+        logging.info(f"{cmd}")
+        subprocess.run(cmd, check=True)
+
+        logging.info(f"dataflow_flex_template_build: {template_gcs_path}")
+        yield template_gcs_path
+        # The template file gets deleted when we delete the bucket.
+
+    @staticmethod
+    def dataflow_flex_template_run(
+        job_name: str,
+        template_path: str,
+        bucket_name: str,
+        parameters: Dict[str, str] = {},
+        project: str = PROJECT,
+        region: str = REGION,
+    ) -> str:
+        import yaml
+
+        # https://cloud.google.com/sdk/gcloud/reference/dataflow/flex-template/run
+        unique_job_name = Utils.hyphen_name(job_name)
+        logging.info(f"dataflow_job_name: {unique_job_name}")
+        cmd = [
+            "gcloud",
+            "dataflow",
+            "flex-template",
+            "run",
+            unique_job_name,
+            f"--template-file-gcs-location={template_path}",
+            f"--project={project}",
+            f"--region={region}",
+            f"--staging-location=gs://{bucket_name}/staging",
+        ] + [
+            f"--parameters={name}={value}"
+            for name, value in {
+                **parameters,
+            }.items()
+        ]
+        logging.info(f"{cmd}")
+
+        try:
+            # The `capture_output` option was added in Python 3.7, so we must
```

Comment:
Okay @davidcavazos in general I'm a fan of your extremely helpful comments but this one is my favorite and I think the most helpful in this file so far 😄